### ISSUE TITLE
Application.GetWindows() raises an error

### ISSUE THREAD

**Issue by uadmin** [gotk3/gotk3#369]
I have been trying to re-write all C examples from GTK docs in Go. I ran into a problem when I tried to re-write an example that uses an `Application` with `glib.APPLICATION_HANDLES_OPEN`.

To reproduce the error:
```golang
package main

import (
  "os"
  "github.com/gotk3/gotk3/gtk"
  "github.com/gotk3/gotk3/glib"
  "fmt"
)

func exampleAppActivate(app *gtk.Application) {
  win, _ := gtk.ApplicationWindowNew(app)
  win.Present()
}

func exampleAppOpen(app *gtk.Application, files unsafe.Pointer, gInt int, hint string) {
  var win *gtk.ApplicationWindow
   windows := app.GetWindows()
  if windows != nil && windows.Length() != 0 {
    win = (windows.Data()).(*gtk.ApplicationWindow)
  } else {
    win, _ = gtk.ApplicationWindowNew(app)
  } 
  // Some code to actually open the file
  win.Present()
}

func main() {
  app, _ := gtk.ApplicationNew("com.github.twistedhardware.gtk.example-4", glib.APPLICATION_HANDLES_OPEN)
  app.Connect("activate", exampleAppActivate)
  app.Connect("open", exampleAppOpen)
  status := app.Run(os.Args)
  fmt.Println(status)
}
```

If you compile the application and run it without passing a file to the application, it runs just fine because it does not invoke `exampleAppOpen`.

To actually reproduce the error:

```bash
$ go build
$ ./example-4 some-file-name
```

Then you will get this error:

```
fatal error: runtime.SetFinalizer: pointer not in allocated block

goroutine 1 [running, locked to thread]:
runtime.throw(0x819904, 0x34)
	/usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc0000a36b8 sp=0xc0000a3688 pc=0x48eff2
runtime.SetFinalizer(0x7a6ea0, 0x0, 0x76a080, 0x81a3a8)
	/usr/local/go/src/runtime/mfinal.go:353 +0x767 fp=0xc0000a37a0 sp=0xc0000a36b8 pc=0x47ba77
github.com/gotk3/gotk3/gtk.(*Application).GetWindows(0xc000084420, 0xc000084540)
	/home/abdullah/go/src/github.com/gotk3/gotk3/gtk/application.go:154 +0x89 fp=0xc0000a37e0 sp=0xc0000a37a0 pc=0x5ae729
main.exampleAppOpen(0xc000084420, 0x2e493b0, 0x1, 0x0, 0x0)
	/home/abdullah/go/src/github.com/twistedhardware/gtk/example-4/main.go:18 +0x2f fp=0xc0000a3818 sp=0xc0000a37e0 pc=0x71911f
runtime.call64(0xc000084450, 0x81ae50, 0xc000084540, 0x2800000028)
	/usr/local/go/src/runtime/asm_amd64.s:520 +0x3b fp=0xc0000a3868 sp=0xc0000a3818 pc=0x4b70cb
reflect.Value.call(0x783fe0, 0x81ae50, 0x13, 0x811597, 0x4, 0xc000090120, 0x4, 0x4, 0x1080900, 0x0, ...)
	/usr/local/go/src/reflect/value.go:447 +0x461 fp=0xc0000a3a88 sp=0xc0000a3868 pc=0x4d9671
reflect.Value.Call(0x783fe0, 0x81ae50, 0x13, 0xc000090120, 0x4, 0x4, 0x1080900, 0x98, 0x476665)
	/usr/local/go/src/reflect/value.go:308 +0xa4 fp=0xc0000a3af0 sp=0xc0000a3a88 pc=0x4d90f4
github.com/gotk3/gotk3/glib.goMarshal(0x2d11530, 0x0, 0x4, 0x7ffdcbdc4990, 0x7ffdcbdc4910, 0x0)
	/home/abdullah/go/src/github.com/gotk3/gotk3/glib/glib.go:233 +0x6f7 fp=0xc0000a3c68 sp=0xc0000a3af0 pc=0x521357
github.com/gotk3/gotk3/glib._cgoexpwrap_8c39c85f7640_goMarshal(0x2d11530, 0x0, 0x4, 0x7ffdcbdc4990, 0x7ffdcbdc4910, 0x0)
	_cgo_gotypes.go:4254 +0x5b fp=0xc0000a3ca8 sp=0xc0000a3c68 pc=0x51ee1b
runtime.call64(0x0, 0x7ffdcbdc46b0, 0x7ffdcbdc4750, 0x30)
	/usr/local/go/src/runtime/asm_amd64.s:520 +0x3b fp=0xc0000a3cf8 sp=0xc0000a3ca8 pc=0x4b70cb
runtime.cgocallbackg1(0x0)
	/usr/local/go/src/runtime/cgocall.go:314 +0x177 fp=0xc0000a3d70 sp=0xc0000a3cf8 pc=0x467c07
runtime.cgocallbackg(0x0)
	/usr/local/go/src/runtime/cgocall.go:191 +0xc5 fp=0xc0000a3dd8 sp=0xc0000a3d70 pc=0x4679f5
runtime.cgocallback_gofunc(0x4678bf, 0x719770, 0xc0000a3e68, 0xc0000a3e58)
	/usr/local/go/src/runtime/asm_amd64.s:773 +0x9b fp=0xc0000a3df8 sp=0xc0000a3dd8 pc=0x4b861b
runtime.asmcgocall(0x719770, 0xc0000a3e68)
	/usr/local/go/src/runtime/asm_amd64.s:620 +0x42 fp=0xc0000a3e00 sp=0xc0000a3df8 pc=0x4b84b2
runtime.cgocall(0x719770, 0xc0000a3e68, 0xc000084390)
	/usr/local/go/src/runtime/cgocall.go:131 +0x7f fp=0xc0000a3e38 sp=0xc0000a3e00 pc=0x4678bf
github.com/gotk3/gotk3/glib._Cfunc_g_application_run(0x2d0f140, 0x2, 0x2d0ef60, 0x0)
	_cgo_gotypes.go:1222 +0x4d fp=0xc0000a3e68 sp=0xc0000a3e38 pc=0x519f8d
github.com/gotk3/gotk3/glib.(*Application).Run.func4(0xc000084390, 0xc00000e0a0, 0x2, 0x2, 0x2d0ef60, 0x2d11530)
	/home/abdullah/go/src/github.com/gotk3/gotk3/glib/application.go:202 +0xa6 fp=0xc0000a3ea8 sp=0xc0000a3e68 pc=0x5289e6
github.com/gotk3/gotk3/glib.(*Application).Run(0xc000084390, 0xc00000e0a0, 0x2, 0x2, 0x0)
	/home/abdullah/go/src/github.com/gotk3/gotk3/glib/application.go:202 +0x13e fp=0xc0000a3f10 sp=0xc0000a3ea8 pc=0x51f6fe
main.main()
	/home/abdullah/go/src/github.com/twistedhardware/gtk/example-4/main.go:32 +0x127 fp=0xc0000a3f98 sp=0xc0000a3f10 pc=0x7192e7
runtime.main()
	/usr/local/go/src/runtime/proc.go:200 +0x20c fp=0xc0000a3fe0 sp=0xc0000a3f98 pc=0x49095c
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc0000a3fe8 sp=0xc0000a3fe0 pc=0x4b8d61
```

**Comment by uadmin** [comment]
I'm drilling down to know where the problem happens, so if you look at the code of the function `Application.GetWindow` you will see this:

```golang
func (v *Application) GetWindows() *glib.List {
	glist := C.gtk_application_get_windows(v.native())
	list := glib.WrapList(uintptr(unsafe.Pointer(glist)))
	// some more code
```

When `list`'s value is `(*glib.List)(nil)` the problem happens. This error does not happen when we have some windows returned from the first call to `gtk_application_get_windows`.

I hope that helps in debugging this.

**Comment by uadmin** [comment]
By the way I found this while I was trying to fix this error:

> Note: the current implementation has a bug. While Go code is permitted to write nil or a C pointer (but not a Go pointer) to C memory, the current implementation may sometimes cause a runtime error if the contents of the C memory appear to be a Go pointer. Therefore, avoid passing uninitialized C memory to Go code if the Go code is going to store pointer values in it. Zero out the memory in C before passing it to Go.

Source: [Golang cgo docs](https://golang.org/cmd/cgo/)

So maybe we want to add an if statement to check if the return is nil before calling `runtime.SetFinalizer` like this:

```golang
// GetWindows is a wrapper around gtk_application_get_windows().
// Returned list is wrapped to return *gtk.Window elements.
func (v *Application) GetWindows() *glib.List {
        glist := C.gtk_application_get_windows(v.native())
        list := glib.WrapList(uintptr(unsafe.Pointer(glist)))
        list.DataWrapper(func(ptr unsafe.Pointer) interface{} {
                return wrapWindow(glib.Take(ptr))
        })
        if list == nil {
                return nil
        }
        runtime.SetFinalizer(list, func(l *glib.List) {
                l.Free()
        })
        return list
}
```

This seems to fix the bug in cgo implementation. If this fix is fine, I can submit a pull request. Any Input?

**Comment by Bios-Marcel** [comment]
As long as nil is a valid value, it seems fine to me. However, can we somehow find out where this fix has to be applied? I doubt it is only this one occurrence.

**Comment by uadmin** [comment]
@Bios-Marcel I bet it has to be applied in a few more places. Maybe a more general solution would be to add it to all wrap functions with a comment that refers to the cgo bug and instructions to remove it when the cgo implementation resolves the bug. I will have to go over `glib` code and see if we can implement it in a way that doesn't have side effects.

I'll check and get back to you.

**Comment by uadmin** [comment]
I just found something even more interesting. This might be a bug in Golang implementation of `runtime.SetFinalizer`. This error can be easily reproduced just by passing any nil pointer to `runtime.SetFinalizer`. To reproduce that:

```golang
package main

import (
  "runtime"
) 

func main() {
  var x *int
  
  runtime.SetFinalizer(x, func(obj *int) {})
}
```

You could add `x=nil` before the `runtime.SetFinalizer` and that also will produce the same exact error.

@Bios-Marcel: I think now a more generic solution (without trying to fix Golang) would be to:

1. Add a wrapper function for `runtime.SetFinalizer` to check if the value is `nil` before calling `runtime.SetFinalizer`.
2. Find all instances of `runtime.SetFinalizer` and add a check before them to insure that we don't pass a `nil` as the first parameter.

Finally I think we should also file a bug report for this in golang repo.

**Comment by Bios-Marcel** [comment]
I wonder what the expected behaviour would be for passing `nil`. What could be considered a bug?
I think the expected behaviour should be returning an error or even panicking, since passing a `nil` value shouldn't really happen ... ever. Maybe someone has already reported this? Now the question is, does wrapping `SetFinalizer` make sense or will it just hide bugs?

**Comment by uadmin** [comment]
OK I finally understand `runtime.SetFinalizer` a little better now.

from: [SetFinalizer docs](https://golang.org/pkg/runtime/#SetFinalizer)

> The argument obj must be a pointer to an object allocated by calling new, by taking the address of a composite literal, or by taking the address of a local variable. The argument finalizer must be a function that takes a single argument to which obj's type can be assigned, and can have arbitrary ignored return values. If either of these is not true, SetFinalizer may abort the program.

So obj can be a the address of a local variable. The problem here is when we set obj to `nil` it doesn't point to any local variable. Let me explain why it is not:

```golang
package main

import (
  "runtime"
) 
  
func main() {
  var a = new(int)
  var b int
  var c *int
  var d *int
  
  var v = 5
  c = &v
 
  runtime.SetFinalizer(a, func(obj *int) {})
  runtime.SetFinalizer(&b, func(obj *int) {})
  runtime.SetFinalizer(c, func(obj *int) {})
  runtime.SetFinalizer(d, func(obj *int) {}) // Only this throws an error
}
```

Now let's see all four scenarios:

1. `a` is a pointer to a int but is "a pointer to a newly allocated zero value of type int". So there is a local variable that `a` points to with value of 0.
2. `b` works fine because it is a local variable of type int and we are passing a pointer to it.
3. `c` also works fine because it points to the address of a local variable `v`.
4. `d` fails because it is just a pointer. It does not point to any local variable. To explain that further, imagine if you want to get the local variable that this pointer is pointing to. You will have to deference it. Deferencing a null pointer is not a valid operation.

So passing a null pointer to `runtime.SetFinalizer` is not a valid operation because this pointer could be pointing that is not pointing to any of the types of variables that are specified by the documentation mentioned above.

So what is the solution for that? I can see a few things that we can do. 

1. Wrapper function for `runtime.SetFinalizer`
2. Add check before `runtime.SetFinalizer`
3. Give up the use of nil pointer in `gtk` and `glib` wrap functions where ever `runtime.SetFinalizer` is used downstream. This is really hard because it is hard to guess what is using your wrap function downstream before when you are just adding it.
4. Try and open an issue in Golang repo with a suggested solution of adding a null pointer check somewhere before they check for memory address ranges [here](https://golang.org/src/runtime/mfinal.go?s=10402:10459#L331) maybe in a way that looks like this:

```golang
	if base == 0 {
		// 0-length objects are okay.
		if e.data == unsafe.Pointer(&zerobase) {
			return
		}

		// Check if the object is a null pointer
		if e.data == nil {
			return
		}

		// More code here that actually checks if the memory address is in range.
```

There are already 3 filed issues but, none of them addresses the issue of null pointer.
https://github.com/golang/go/issues/22350
https://github.com/golang/go/issues/17311
https://github.com/golang/go/issues/16081

@Bios-Marcel What do you think? I know none of these solutions look good or easy to implement and I hope someone smarter will show up and find a better solutions.

**Comment by Bios-Marcel** [comment]
Well, like already mentioned, I think `runtime#SetFinalizer` should return an error. It should always be the APIs job to keep the developer from doing stupid things. Assigning a finalizer to `nil` pointer seems very stupid to me.

However, in our case we just seem to be making an assumption that `glib#WrapList` can handle values that are possibly nil. Or we didn't even think this could be nil. And `nil` seems to be a correct return value from `gtk_application_get_windows`, we should just not try to wrap the result if it is `nil`. So, what I'd do is to add a nil check before calling `glib#WrapList` and clearify in the documentation, that it can't handle null values. Yes, it might be tedious to do this everywhere, but I think it would be the correct way. We also couldn't really add a new error return value to `glib#WrapList`, since we'd be breaking backwards compatibility.

**Comment by uadmin** [comment]
@Bios-Marcel Sorry, I forgot to ask if you need help with that. I could start by looking for anywhere we use `runtime.SetFinalizer` and check upstream code for a `nill` value.

BTW, I don't want to overwhelm this you or the project. So I have one dedicated developer who is assigned GTK related work who will be happy to do any work that is related to this repo.

**Comment by Bios-Marcel** [comment]
We'd surely be glad to have more contributors. I currently don't have much time myself, which is why I mostly only fix super-simple and tiny things, answer issues and merge PRs.

**Comment by uadmin** [comment]
OK. I'll have that work done and submit a pull request for it. I'll refer to this issue in the pull request.

**Comment by andre-hub** [comment]
some methods include a check so that certain parameters are not further processed with zero. according to the principle of fast break and error message create.

e.g glib#WrapList

**Comment by MJacred** [comment]
has been fixed with #539.  
Issue can be closed
